### PR TITLE
E2E tests Refactoring

### PR DESCRIPTION
**Pull request by palexster**
# Description

This PR starts the migration of E2E tests to Ginkgo/gomega framework.

### CODE REVIEW

File: test/e2e/unjoin_e2e/unpeering_test.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+			readyPods, notReadyPods, err := util.ArePodsUp(testContext.Clusters[1].Client, testContext.Namespace)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(notReadyPods).Should(BeZero())
+			Expect(len(readyPods)).Should(BeNumerically(">", 0))
+		},
+		)
+	})
+})
+
+func NoPods(clientset *kubernetes.Clientset, namespace string) (bool, error) {
```
Comment (giorio94) [inline test/e2e/unjoin_e2e/unpeering_test.go:47]: The boolean value seems redundant to me (or the error one)

File: test/e2e/unjoin_e2e/unpeering_test.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	pods, err := clientset.CoreV1().Pods(namespace).List(context2.TODO(), metav1.ListOptions{})
+	if err != nil {
+		klog.Error(err)
+		return false, nil
```
Comment (giorio94) [inline test/e2e/unjoin_e2e/unpeering_test.go:51]:
```suggestion
		return false, err
```

File: test/e2e/unjoin_e2e/unpeering_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	}
+	if len(pods.Items) > 0 {
+		return false, fmt.Errorf("There are still running pods in Liqo namespace")
+	}
+	return true, nil
+}
+
+func NoJoined(clientset *kubernetes.Clientset) (bool, error) {
```
Comment (giorio94) [inline test/e2e/unjoin_e2e/unpeering_test.go:59]: Same here about return values redundancy.

**Review by giorio94 — COMMENTED** [review]
Two general comments:
1. The microservices app test appears to me nondeterministic, since no affinities are specified and the actual placement is left to the scheduler. Hence, all pods could be scheduled locally (and you would be testing k8s) and, even in case of errors, they would be considered glitches as non reproducible.  
2. Contexts should be propagated, and not created on-demand, otherwise they become useless.

More comments inline (sorry for their number). Feel free to discard them if they don't make sense to you

File: test/e2e/testutils/consts.go [hunk 4]
```diff
@@ -0,0 +1,8 @@
+// Package testutils encapsulates all methods and constants to perform E2E tests
+package testutils
+
+// LiqoTestNamespaceLabels is a set of label that has to be attached to test namespace to simplify garbage collection.
+var LiqoTestNamespaceLabels = map[string]string{
+	"liqo.io/testing-namespace": "true",
+	"liqo.io/enabled":           "true",
```
Comment (aleoli) [inline test/e2e/testutils/consts.go:7]: We should use a const defined in the consts package for this value

File: test/e2e/testutils/net/net.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
+package net
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/liqotech/liqo/test/e2e/testutils/util"
```
Comment (aleoli) [inline test/e2e/testutils/net/net.go:8]: This should be grouped with the last import group

File: test/e2e/testutils/util/nodes.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	klog.Info(len(nodes.Items))
+	if err != nil {
+		klog.Error(err)
+		return false
+	}
+	node, err := homeClusterClient.CoreV1().Nodes().Get(context.TODO(), "liqo-"+id.Data["cluster-id"], metav1.GetOptions{})
+	if err != nil {
+		klog.Error(err)
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (aleoli) [reply test/e2e/testutils/util/nodes.go:41]: Yes, if the intention is to get the virtual nodes, we can list them by label, instead of getting it by name

**Comment by palexster** [comment]
Hi @giorio94 and @aleoli, thank you for your reviews!

> Two general comments:
> 
>     1. The microservices app test appears to me nondeterministic, since no affinities are specified and the actual placement is left to the scheduler. Hence, all pods could be scheduled locally (and you would be testing k8s) and, even in case of errors, they would be considered glitches as non reproducible.

100% agree. Moreover, the check is limited and performed only on the homepage. Since this limitation is not introduced by this PR, for sake of brevity I would let this PR as it is and prepare a follow-up PR with a new YAML and checks that touch also the other microservices. 

The idea of this first PR is to have a slightly more consolidated code-base in order to simplify new test additions to the E2E suite.

> 
>     2. Contexts should be propagated, and not created on-demand, otherwise they become useless.

Ok, no problem with that. Personal question: is it not really clear to me what can contexts bring in for this test scenario.

> 
> 
> More comments inline (sorry for their number). Feel free to discard them if they don't make sense to you

**Comment by giorio94** [comment]
> Hi @giorio94 and @aleoli, thank you for your reviews!
> 
> 
> 
> > Two general comments:
> 
> > 
> 
> >     1. The microservices app test appears to me nondeterministic, since no affinities are specified and the actual placement is left to the scheduler. Hence, all pods could be scheduled locally (and you would be testing k8s) and, even in case of errors, they would be considered glitches as non reproducible.
> 
> 
> 
> 100% agree. Moreover, the check is limited and performed only on the homepage. Since this limitation is not introduced by this PR, for sake of brevity I would let this PR as it is and prepare a follow-up PR with a new YAML and checks that touch also the other microservices. 
> 
> 
> 
> The idea of this first PR is to have a slightly more consolidated code-base in order to simplify new test additions to the E2E suite.
> 
> 

Definitely fine from my point of view. Indeed, this is already a quite big PR.

> 
> > 
> 
> >     2. Contexts should be propagated, and not created on-demand, otherwise they become useless.
> 
> 
> 
> Ok, no problem with that. Personal question: is it not really clear to me what can contexts bring in for this test scenario.
> 
> 

I think they should allow for a more ordinated termination in case of aborts, to prevent requests hanging (although I am not sure how/whether they are supported by Ginkgo), but I agree they are not that important in case of tests. Still, since some utility functions may be of larger use, I feel it would be better to support them for the sake of generality. Just personal opinion though.

> 
> > 
> 
> > 
> 
> > More comments inline (sorry for their number). Feel free to discard them if they don't make sense to you
> 
> 
> 
>

File: test/e2e/peering_e2e/basic_test.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+				Entry("Check Pod to Pod connectivity from cluster 1", testContext.Clusters[0], testContext.Clusters[1], namespace),
+				Entry("Check Pod to Pod connectivity from cluster 2", testContext.Clusters[1], testContext.Clusters[0], namespace),
+			)
+		})
+
+		Context("E2E Testing with Online Boutique", func() {
+
+			options := k8s.NewKubectlOptions("", testContext.Clusters[0].KubeconfigPath, microservices.TestNamespaceName)
+			defer GinkgoRecover()
```
[the remark this answers is not in the document]
Comment (palexster) [reply test/e2e/peering_e2e/basic_test.go:97]: As suggested by the failure log, this call prevents the test failure from being fatal and continues other test executions.

File: test/e2e/testutils/net/net.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+func CheckNodeToPortConnectivity(homeClusterClient kubernetes.Interface, homeClusterID string, nodePortValue int) error {
+	localNodes, err := util.GetNodes(homeClusterClient, homeClusterID, labelSelectorNodes)
+	if err != nil {
+		return err
+	}
+	return util.TriggerCheckNodeConnectivity(localNodes, command, nodePortValue)
+}
+
+// CheckCorrectScheduling checks that Pods are scheduled to the expected nodes.
+func CheckCorrectScheduling(homeClient, foreignClient kubernetes.Interface, homeID, foreignID string) error {
```
[the remark this answers is not in the document]
Comment (palexster) [reply test/e2e/testutils/net/net.go:61]: It was present in the old test. 100% Agree to drop it

File: test/e2e/testutils/microservices/deploy_app.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	if err := k8s.KubectlApplyE(t, options, kubeResourcePath); err != nil {
+		return err
+	}
+	return nil
+}
+
+// WaitDemoApp waits until each service of the application has ready endpoints. It fails if this does not happen
+// within the timeout (retries*sleepBetweenRetries).
+func WaitDemoApp(t ginkgo.GinkgoTInterface, options *k8s.KubectlOptions) {
+	pods := k8s.ListPods(t, options, metav1.ListOptions{})
```
[the remark this answers is not in the document]
Comment (palexster) [reply test/e2e/testutils/microservices/deploy_app.go:45]: 100% agree. Would add this refactoring to a new PR to not change too much in this one.

**Comment by palexster** [comment]
-rebase

File: test/e2e/testutils/util/nodes.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	virtualNodes, err := homeClusterClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(nodeLabel).String(),
+	})
+	if err != nil {
+		klog.Error(err)
+		return false
+	}
+	for index := range virtualNodes.Items {
+		for _, condition := range virtualNodes.Items[index].Status.Conditions {
+			if condition.Type == "Ready" {
```
Comment (aleoli) [inline test/e2e/testutils/util/nodes.go:40]:
```suggestion
			if condition.Type == v1.NodeReady {
```

File: test/e2e/testutils/util/nodes.go [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+				if string(condition.Status) == "False" {
+					return false
+				}
+			} else {
+				if string(condition.Status) == "True" {
```
Comment (aleoli) [inline test/e2e/testutils/util/nodes.go:45]:
```suggestion
				if string(condition.Status) == v1.ConditionTrue {
```

File: test/e2e/testutils/util/nodes.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+		LabelSelector: labels.SelectorFromSet(nodeLabel).String(),
+	})
+	if err != nil {
+		klog.Error(err)
+		return false
+	}
+	for index := range virtualNodes.Items {
+		for _, condition := range virtualNodes.Items[index].Status.Conditions {
+			if condition.Type == "Ready" {
+				if string(condition.Status) == "False" {
```
Comment (aleoli) [inline test/e2e/testutils/util/nodes.go:41]:
```suggestion
				if string(condition.Status) == v1.ConditionFalse {
```

File: test/e2e/testutils/net/pod.go [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+// CheckTesterPods retrieves the netTest pods and returns true if all the pods are up and ready.
+func CheckTesterPods(ctx context.Context, homeClient, foreignClient kubernetes.Interface, homeClusterID string) bool {
+	reflectedNamespace := TestNamespaceName + "-" + homeClusterID
+	return util.IsPodUp(ctx, homeClient, TestNamespaceName, podTesterLocalCl, true) &&
+		util.IsPodUp(ctx, homeClient, TestNamespaceName, podTesterRemoteCl, true) &&
+		util.IsPodUp(ctx, foreignClient, reflectedNamespace, podTesterRemoteCl, false)
+}
+
+// ForgeTesterPod deploys the Remote pod of the test.
+func ForgeTesterPod(image, podName, namespace string, isRemote bool) *v1.Pod {
```
Comment (aleoli) [inline test/e2e/testutils/net/pod.go:47]: does it need to be an exported function?

File: test/e2e/unjoin_e2e/unpeering_test.go [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
```
Comment (giorio94) [inline test/e2e/unjoin_e2e/unpeering_test.go:12]:
```suggestion
	"k8s.io/klog/v2"
```

**Review by aleoli — APPROVED** [review]

**Review by giorio94 — APPROVED** [review]
/lgtm

File: test/e2e/peering_e2e/basic_test.go [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+					Expect(notReadyPods).To(BeNil())
+					Expect(len(readyPods)).Should(BeNumerically(">", 0))
+				},
+				Entry("Pods UP on cluster 1", testContext.Clusters[0], namespace),
+				Entry("Pods UP on cluster 2", testContext.Clusters[1], namespace),
+			)
+
+			DescribeTable("Liqo Virtual Nodes are ready",
+
+				func(homeCluster tester.ClusterContext, foreignCluster tester.ClusterContext, namespace string) {
```
Comment (Andreagit97) [inline test/e2e/peering_e2e/basic_test.go:52]: the second parameter `foreignCluster ` may not be needed here.

File: pkg/consts/clusterid.go [hunk 16]
```diff
@@ -1,9 +1,14 @@
...
-	ConfigMapKey           = "cluster-id"
-	ClusterIDLabelName     = "clusterID"
+	// ClusterIDConfigMapName is the name of the configmap where the cluster-id is stored.
+	ClusterIDConfigMapName = "cluster-id"
+	// MasterLabel contains the label used to identify the master nodes.
+	MasterLabel = "node-role.kubernetes.io/master"
+	// ServiceAccountNamespacePath contains the path where the namespace is stored in the serviceaccount volume mount.
+	ServiceAccountNamespacePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+	// ClusterIDLabelName is the name of the label key to use with Cluster ID.
+	ClusterIDLabelName = "cluster-id"
```
[the remark this answers is not in the document]
Comment (palexster) [reply pkg/consts/clusterid.go:11]: No, it's a mistake. Thanks!

**Pull request merged**